File: backend/app/api/goals.py
"""Goals API — router fino (A6e.4 · ADR-101 R15/R16 · ADR-073).

Endpoints sob ``/workspaces/{workspace_id}/goals/...`` delegam a use
cases em :mod:`backend.app.application.goal`. Erros de domínio traduzidos
para HTTP por handlers globais em ``main.py``.

4 tipos versionados append-only: IF, aportes, dolarização, alocação.
"""

from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from backend.app.application.goal import (
    compute_alocacao_projection,
    compute_aporte_projection,
    compute_dolar_projection,
    compute_if_projection,
    create_if_goal_version,
    create_typed_goal_version,
    get_active_if_goal,
    get_active_typed_goal,
    list_if_goal_versions,
    list_typed_goal_versions,
)
from backend.app.application.goal._author_enrichment import (
    attach_author_name,
    resolve_author_names,
)
from backend.app.core.database import get_db
from backend.app.core.deps import get_current_user
from backend.app.core.tenancy import get_current_workspace, require_write_role
from backend.app.models.user import User
from backend.app.models.workspace import Workspace
from backend.app.repositories.goal_repository import GoalRepository
from backend.app.schemas.dto.goal import (
    AlocacaoGoalComputeRequest,
    AlocacaoGoalComputeResponse,
    AlocacaoGoalHistoryResponse,
    AlocacaoGoalResponse,
    AlocacaoGoalUpsertCommand,
    AporteGoalComputeRequest,
    AporteGoalComputeResponse,
    AporteGoalHistoryResponse,
    AporteGoalResponse,
    AporteGoalUpsertCommand,
    DolarGoalComputeRequest,
    DolarGoalComputeResponse,
    DolarGoalHistoryResponse,
    DolarGoalResponse,
    DolarGoalUpsertCommand,
    IFGoalComputeRequest,
    IFGoalComputeResponse,
    IFGoalHistoryResponse,
    IFGoalResponse,
    IFGoalUpsertCommand,
)
from backend.app.schemas.task import TaskFilters, TaskListResponse, TaskResponse
from backend.app.services import task_service
from backend.app.services.goal_service import get_latest_report_patrimonio_liquido

router = APIRouter(prefix="/workspaces/{workspace_id}/goals", tags=["goals"])


def _get_repo(db: AsyncSession = Depends(get_db)) -> GoalRepository:
    return GoalRepository(db)


async def _history_typed(
    goal_type: str, workspace_id: str, *, repo: GoalRepository, db: AsyncSession
) -> list[BaseModel]:
    goals = await repo.list_by_workspace_and_type(workspace_id, goal_type)
    names = await resolve_author_names(
        {g.created_by for g in goals if g.created_by}, db=db
    )
    return await list_typed_goal_versions(
        workspace_id, goal_type, repo=repo, author_names=names
    )


async def _read_typed(
    goal_type: str, workspace_id: str, *, repo: GoalRepository, db: AsyncSession
) -> BaseModel:
    resp = await get_active_typed_goal(workspace_id, goal_type, repo=repo)
    return await attach_author_name(resp, db=db)


async def _write_typed(
    goal_type: str, inputs: BaseModel, notes: Optional[str], *,
    workspace: Workspace, user: User, repo: GoalRepository, db: AsyncSession,
) -> BaseModel:
    resp = await create_typed_goal_version(
        goal_type, inputs, notes,
        workspace_id=workspace.id, created_by=user.id, repo=repo,
        created_by_name=user.full_name,
    )
    await db.commit()
    return resp


# INDEPENDENCIA_FINANCEIRA (F8.1) ════════════════════════════════════════


@router.post("/if/compute", response_model=IFGoalComputeResponse,
             summary="Dry-run: calcula derivados sem persistir")
async def compute_if_goal(
    body: IFGoalComputeRequest,
    workspace: Workspace = Depends(get_current_workspace),
    user: User = Depends(get_current_user),
) -> IFGoalComputeResponse:
    return compute_if_projection(body)


@router.get("/if", response_model=IFGoalResponse,
            summary="Retorna a meta IF vigente do workspace")
async def get_if_goal(
    workspace: Workspace = Depends(get_current_workspace),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> IFGoalResponse:
    patrimonio = await get_latest_report_patrimonio_liquido(workspace.id, db=db)
    resp = await get_active_if_goal(
        workspace.id, repo=repo, patrimonio_atual_brl=patrimonio
    )
    return await attach_author_name(resp, db=db)  # type: ignore[return-value]


@router.get("/if/history", response_model=IFGoalHistoryResponse,
            summary="Histórico completo de versões da meta IF")
async def get_if_goal_history(
    workspace: Workspace = Depends(get_current_workspace),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> IFGoalHistoryResponse:
    goals = await repo.list_by_workspace_and_type(
        workspace.id, "INDEPENDENCIA_FINANCEIRA"
    )
    names = await resolve_author_names(
        {g.created_by for g in goals if g.created_by}, db=db
    )
    return await list_if_goal_versions(workspace.id, repo=repo, author_names=names)


@router.put("/if", response_model=IFGoalResponse,
            summary="Cria nova versão da meta IF (fecha a anterior)",
            dependencies=[Depends(require_write_role)])
async def upsert_if_goal(
    body: IFGoalUpsertCommand,
    workspace: Workspace = Depends(get_current_workspace),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> IFGoalResponse:
    patrimonio = await get_latest_report_patrimonio_liquido(workspace.id, db=db)
    resp = await create_if_goal_version(
        body, workspace_id=workspace.id, created_by=user.id, repo=repo,
        patrimonio_atual_brl=patrimonio, created_by_name=user.full_name,
    )
    await db.commit()
    return resp


# APORTE_MENSAL (F8.5) ═══════════════════════════════════════════════════


@router.post("/aportes/compute", response_model=AporteGoalComputeResponse,
             summary="Dry-run: calcula derivados de aportes")
async def compute_aporte_goal(
    body: AporteGoalComputeRequest,
    workspace: Workspace = Depends(get_current_workspace),
    user: User = Depends(get_current_user),
) -> AporteGoalComputeResponse:
    return compute_aporte_projection(body)


@router.get("/aportes", response_model=AporteGoalResponse,
            summary="Meta de aportes vigente")
async def get_aporte_goal(
    workspace: Workspace = Depends(get_current_workspace),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> AporteGoalResponse:
    return await _read_typed("APORTE_MENSAL", workspace.id, repo=repo, db=db)  # type: ignore[return-value]


@router.get("/aportes/history", response_model=AporteGoalHistoryResponse,
            summary="Histórico de versões da meta de aportes")
async def get_aporte_goal_history(
    workspace: Workspace = Depends(get_current_workspace),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> AporteGoalHistoryResponse:
    responses = await _history_typed("APORTE_MENSAL", workspace.id, repo=repo, db=db)
    return AporteGoalHistoryResponse(goals=responses, total=len(responses))  # type: ignore[arg-type]


@router.put("/aportes", response_model=AporteGoalResponse,
            summary="Cria nova versão da meta de aportes",
            dependencies=[Depends(require_write_role)])
async def upsert_aporte_goal(
    body: AporteGoalUpsertCommand,
    workspace: Workspace = Depends(get_current_workspace),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> AporteGoalResponse:
    return await _write_typed(  # type: ignore[return-value]
        "APORTE_MENSAL", body.inputs, body.notes,
        workspace=workspace, user=user, repo=repo, db=db,
    )


# DOLARIZACAO (F8.5) ═════════════════════════════════════════════════════


@router.post("/dolarizacao/compute", response_model=DolarGoalComputeResponse,
             summary="Dry-run: calcula derivados de dolarização")
async def compute_dolar_goal(
    body: DolarGoalComputeRequest,
    workspace: Workspace = Depends(get_current_workspace),
    user: User = Depends(get_current_user),
) -> DolarGoalComputeResponse:
    return compute_dolar_projection(body)


@router.get("/dolarizacao", response_model=DolarGoalResponse,
            summary="Meta de dolarização vigente")
async def get_dolar_goal(
    workspace: Workspace = Depends(get_current_workspace),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> DolarGoalResponse:
    return await _read_typed("DOLARIZACAO", workspace.id, repo=repo, db=db)  # type: ignore[return-value]


@router.get("/dolarizacao/history", response_model=DolarGoalHistoryResponse,
            summary="Histórico de versões da meta de dolarização")
async def get_dolar_goal_history(
    workspace: Workspace = Depends(get_current_workspace),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> DolarGoalHistoryResponse:
    responses = await _history_typed("DOLARIZACAO", workspace.id, repo=repo, db=db)
    return DolarGoalHistoryResponse(goals=responses, total=len(responses))  # type: ignore[arg-type]


@router.put("/dolarizacao", response_model=DolarGoalResponse,
            summary="Cria nova versão da meta de dolarização",
            dependencies=[Depends(require_write_role)])
async def upsert_dolar_goal(
    body: DolarGoalUpsertCommand,
    workspace: Workspace = Depends(get_current_workspace),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> DolarGoalResponse:
    return await _write_typed(  # type: ignore[return-value]
        "DOLARIZACAO", body.inputs, body.notes,
        workspace=workspace, user=user, repo=repo, db=db,
    )


# ALOCACAO_ALVO (F8.5) ═══════════════════════════════════════════════════


@router.post("/alocacao/compute", response_model=AlocacaoGoalComputeResponse,
             summary="Dry-run: valida alocação-alvo")
async def compute_alocacao_goal(
    body: AlocacaoGoalComputeRequest,
    workspace: Workspace = Depends(get_current_workspace),
    user: User = Depends(get_current_user),
) -> AlocacaoGoalComputeResponse:
    return compute_alocacao_projection(body)


@router.get("/alocacao", response_model=AlocacaoGoalResponse,
            summary="Alocação-alvo vigente")
async def get_alocacao_goal(
    workspace: Workspace = Depends(get_current_workspace),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> AlocacaoGoalResponse:
    return await _read_typed("ALOCACAO_ALVO", workspace.id, repo=repo, db=db)  # type: ignore[return-value]


@router.get("/alocacao/history", response_model=AlocacaoGoalHistoryResponse,
            summary="Histórico de versões da alocação-alvo")
async def get_alocacao_goal_history(
    workspace: Workspace = Depends(get_current_workspace),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> AlocacaoGoalHistoryResponse:
    responses = await _history_typed("ALOCACAO_ALVO", workspace.id, repo=repo, db=db)
    return AlocacaoGoalHistoryResponse(goals=responses, total=len(responses))  # type: ignore[arg-type]


@router.put("/alocacao", response_model=AlocacaoGoalResponse,
            summary="Cria nova versão da alocação-alvo",
            dependencies=[Depends(require_write_role)])
async def upsert_alocacao_goal(
    body: AlocacaoGoalUpsertCommand,
    workspace: Workspace = Depends(get_current_workspace),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    repo: GoalRepository = Depends(_get_repo),
) -> AlocacaoGoalResponse:
    return await _write_typed(  # type: ignore[return-value]
        "ALOCACAO_ALVO", body.inputs, body.notes,
        workspace=workspace, user=user, repo=repo, db=db,
    )


# Tasks linked to a goal (cross-aggregate read) ══════════════════════════


@router.get("/{goal_id}/tasks", response_model=TaskListResponse,
            summary="Tarefas vinculadas a esta meta")
async def list_tasks_for_goal(
    goal_id: str,
    workspace: Workspace = Depends(get_current_workspace),
    db: AsyncSession = Depends(get_db),
    include_done: bool = False,
) -> TaskListResponse:
    filters = TaskFilters(related_goal_id=goal_id, include_done=include_done)
    tasks = await task_service.list_tasks(workspace.id, filters, db=db)
    return TaskListResponse(
        tasks=[TaskResponse.model_validate(t) for t in tasks],
        total=len(tasks),
    )

File: frontend/src/lib/api/goals.ts
import { apiFetch } from "./core";

// ═══════════════════════════════════════════════════════════════════════
// Goals — Meta IF (ADR-073, F8.1)
// Padrão F8+: endpoints escopados por {API_BASE}/workspaces/{ws_id}/...
// ═══════════════════════════════════════════════════════════════════════

export interface IFGoalInputs {
  renda_passiva_mensal_brl: number;
  trs_pct: number;
  retorno_real_anual_pct: number;
  horizonte_anos: number;
  taxa_retirada_conservadora_pct?: number;
}

export interface IFGoalDerived {
  if_meta_brl: number;
  /** Aporte assumindo patrimônio inicial zero (baseline; persistido). */
  aporte_necessario_mensal_brl: number;
  if_meta_conservadora_brl: number;
  /** Quando há patrimônio de referência (ex.: último relatório), aporte para fechar o gap. */
  aporte_mensal_com_patrimonio_atual_brl?: number | null;
  patrimonio_atual_utilizado_brl?: number | null;
}

/** Aporte a exibir: ajustado ao patrimônio conhecido, senão baseline. */
export function ifMonthlyContributionDisplay(d: IFGoalDerived): number {
  return (
    d.aporte_mensal_com_patrimonio_atual_brl ?? d.aporte_necessario_mensal_brl
  );
}

export interface IFGoalResponse {
  id: string;
  workspace_id: string;
  type: "INDEPENDENCIA_FINANCEIRA";
  /** Schema canônico (`config/schemas/goal.if.schema.json`). */
  meta_version: number;
  inputs: IFGoalInputs;
  derived: IFGoalDerived;
  effective_from: string;
  effective_to: string | null;
  is_template: boolean;
  notes: string | null;
  created_by: string | null;
  /** F9 · nome humano do autor (join com users.full_name). */
  created_by_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface IFGoalHistoryResponse {
  goals: IFGoalResponse[];
  total: number;
}

export interface IFGoalComputeRequest {
  inputs: IFGoalInputs;
  patrimonio_atual_brl?: number;
}

export interface IFGoalComputeResponse {
  derived: IFGoalDerived;
  percentual_conquistado: number | null;
  faltante_brl: number | null;
}

export async function computeIFGoal(
  workspaceId: string,
  body: IFGoalComputeRequest
): Promise<IFGoalComputeResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/if/compute`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function getIFGoal(
  workspaceId: string
): Promise<IFGoalResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/if`);
}

export async function getIFGoalHistory(
  workspaceId: string
): Promise<IFGoalHistoryResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/if/history`);
}

export async function upsertIFGoal(
  workspaceId: string,
  inputs: IFGoalInputs,
  notes?: string
): Promise<IFGoalResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/if`, {
    method: "PUT",
    body: JSON.stringify({ inputs, notes }),
  });
}

// ═══════════════════════════════════════════════════════════════════════
// Goals — Aportes Mensais (F8.5)
// ═══════════════════════════════════════════════════════════════════════

export interface AporteGoalInputs {
  meta_aporte_mensal_brl: number;
  dia_aporte: number;
  periodo_inicio?: string;
  distribuicao?: Record<string, number>;
}

export interface AporteGoalDerived {
  aporte_anual_brl: number;
  distribuicao_pct: Record<string, number>;
}

export interface AporteGoalResponse {
  id: string;
  workspace_id: string;
  type: "APORTE_MENSAL";
  meta_version: number;
  inputs: AporteGoalInputs;
  derived: AporteGoalDerived;
  effective_from: string;
  effective_to: string | null;
  is_template: boolean;
  notes: string | null;
  created_by: string | null;
  created_by_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface AporteGoalHistoryResponse {
  goals: AporteGoalResponse[];
  total: number;
}

export interface AporteGoalComputeResponse {
  derived: AporteGoalDerived;
}

export async function computeAporteGoal(
  workspaceId: string,
  inputs: AporteGoalInputs
): Promise<AporteGoalComputeResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/aportes/compute`, {
    method: "POST",
    body: JSON.stringify({ inputs }),
  });
}

export async function getAporteGoal(
  workspaceId: string
): Promise<AporteGoalResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/aportes`);
}

export async function getAporteGoalHistory(
  workspaceId: string
): Promise<AporteGoalHistoryResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/aportes/history`);
}

export async function upsertAporteGoal(
  workspaceId: string,
  inputs: AporteGoalInputs,
  notes?: string
): Promise<AporteGoalResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/aportes`, {
    method: "PUT",
    body: JSON.stringify({ inputs, notes }),
  });
}

// ═══════════════════════════════════════════════════════════════════════
// Goals — Dolarização (F8.5)
// ═══════════════════════════════════════════════════════════════════════

export interface DolarGoalInputs {
  meta_usd: number;
  aporte_mensal_brl: number;
}

export interface DolarGoalDerived {
  horizonte_estimado_meses: number;
}

export interface DolarGoalResponse {
  id: string;
  workspace_id: string;
  type: "DOLARIZACAO";
  meta_version: number;
  inputs: DolarGoalInputs;
  derived: DolarGoalDerived;
  effective_from: string;
  effective_to: string | null;
  is_template: boolean;
  notes: string | null;
  created_by: string | null;
  created_by_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface DolarGoalHistoryResponse {
  goals: DolarGoalResponse[];
  total: number;
}

export interface DolarGoalComputeResponse {
  derived: DolarGoalDerived;
  cambio_utilizado: number;
}

export async function computeDolarGoal(
  workspaceId: string,
  inputs: DolarGoalInputs,
  cambio_brl_usd?: number
): Promise<DolarGoalComputeResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/dolarizacao/compute`, {
    method: "POST",
    body: JSON.stringify({ inputs, cambio_brl_usd }),
  });
}

export async function getDolarGoal(
  workspaceId: string
): Promise<DolarGoalResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/dolarizacao`);
}

export async function getDolarGoalHistory(
  workspaceId: string
): Promise<DolarGoalHistoryResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/dolarizacao/history`);
}

export async function upsertDolarGoal(
  workspaceId: string,
  inputs: DolarGoalInputs,
  notes?: string
): Promise<DolarGoalResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/dolarizacao`, {
    method: "PUT",
    body: JSON.stringify({ inputs, notes }),
  });
}

// ═══════════════════════════════════════════════════════════════════════
// Goals — Alocação-Alvo (F8.5)
// ═══════════════════════════════════════════════════════════════════════

export interface AlocacaoGoalInputs {
  renda_fixa_pct: number;
  acoes_pct: number;
  imoveis_reits_pct: number;
  liquidez_usd_pct: number;
  instrumentos_rf?: string;
  instrumentos_rv?: string;
  rebalanceamento?: string;
}

export interface AlocacaoGoalDerived {
  soma_percentuais: number;
}

export interface AlocacaoGoalResponse {
  id: string;
  workspace_id: string;
  type: "ALOCACAO_ALVO";
  meta_version: number;
  inputs: AlocacaoGoalInputs;
  derived: AlocacaoGoalDerived;
  effective_from: string;
  effective_to: string | null;
  is_template: boolean;
  notes: string | null;
  created_by: string | null;
  created_by_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface AlocacaoGoalHistoryResponse {
  goals: AlocacaoGoalResponse[];
  total: number;
}

export interface AlocacaoGoalComputeResponse {
  derived: AlocacaoGoalDerived;
  valido: boolean;
}

export async function computeAlocacaoGoal(
  workspaceId: string,
  inputs: AlocacaoGoalInputs
): Promise<AlocacaoGoalComputeResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/alocacao/compute`, {
    method: "POST",
    body: JSON.stringify({ inputs }),
  });
}

export async function getAlocacaoGoal(
  workspaceId: string
): Promise<AlocacaoGoalResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/alocacao`);
}

export async function getAlocacaoGoalHistory(
  workspaceId: string
): Promise<AlocacaoGoalHistoryResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/alocacao/history`);
}

export async function upsertAlocacaoGoal(
  workspaceId: string,
  inputs: AlocacaoGoalInputs,
  notes?: string
): Promise<AlocacaoGoalResponse> {
  return apiFetch(`/workspaces/${workspaceId}/goals/alocacao`, {
    method: "PUT",
    body: JSON.stringify({ inputs, notes }),
  });
}
